src/syntax/grammar.ml

```diff
@@ -127,7 +127,7 @@ let parsing_macro_cond = ref false
 let rec parse_file s =
 	last_doc := None;
 	match%parser s with
-	| [ (Kwd Package,_); parse_package as pack; [%s s] ] ->
+	| [ (Kwd Package,_); parse_package as pack ] ->
```

Comment:
@Simn I also considered removing all these s bindings, as ppx_parser didn't originally have this feature. The reason I opted to keep the [%s s] in cases like this was because this `match` is matching on the new state of the stream after `Kwd Package, ... etc` have been popped, not the stream as it was originally passed into the function. This doesn't change the functionality since the Stream objects are mutable, however, I think the code is more understandable (and more ocaml-y) when it's clear that `s` refers to the stream with everything popped.

I would be for a change to refactor the parser to remove these instances of `[%s s]`, however, I don't think this is the correct way to refactor it. I think `match%parser s with` should be removed along with it.

Reply:
Maybe I'm just too dumb to use it properly, but every time I try to deal with OCaml's `function` feature I end up regretting it. As soon as you want to add something (like the `last_doc := None` here) you have to awkwardly refactor thing to use an explicit `match` after all.

I disagree regarding readability too, because nobody is going to have any intuition wtf `[%s s]` is and why it appears in some places and not in others. Making it look more ocaml-y is also not something I'm particularly interested in, because the mutable nature of streams is already pretty un-ocaml-y.

However, I'm interested to see how you would rewrite this without `match%parser s with`.